### ISSUE TITLE
Firefox performance when pasting/loading html

### ISSUE THREAD

**Issue by RobAley** [quilljs/quill#1130]
Firefox is extremely sluggish when pasting in large HTML contents or creating quill on an element with large HTML contents, compared to Chrome/Chromium.

**Steps for Reproduction**

1. Visit http://shakespeare.mit.edu/hamlet/full.html
2. Ctrl-A, Ctrl-C to copy the contents of the page
3. Visit the main Quill Website Interactive Playground.
4. Click in the first codepen Quill editor, and Ctrl-V paste

**Expected behavior**:

The contents are pasted into the quill editor and ready to edit within a couple of seconds.

**Actual behavior**:

The paste takes ~36 seconds in firefox, including 2 "script has become unresponsive" warnings. On Chromium/Chrome it takes ~4 seconds.

In all browsers editing is fine/performant once the paste finishes.

**Platforms**:

Tested on 
Linux (Ubuntu 16.04) Firefox 50.0 Chromium 53.0
Windows (Windows 10) Firefox 49.0 Chrome 54.0
Both on a core i5-6400 16Gb Ram. (results are similar on my older lowerspec laptop).

**Version**:

The latest in use in the playground.

**Comment by jhchen** [comment]
Performance can always be improved, but it should be noted that Hamlet is 126 pages in 12px font in Word.

**Comment by RobAley** [comment]
Indeed, I picked Hamlet as an extreme example so that anyone testing on the fastest of machines would still see the problem, but it is also noticeable on much shorter "long" documents on slower machines.

I think perhaps my intention was to illustrate that the performance in Firefox was _so_ much worse than Chrome(ium), rather than a general performance issue with Quill. Chrome loading Hamlet in 4 seconds shows that there's no fundamental problem with Quill opening such long documents (and 4 seconds is more than reasonable for such a long doc), but something is drastically slowing Firefox on the same task. Docs that open in the blink of an eye on Chrome can still take a noticeable time on Firefox.

While it's likely to be Firefox's "fault" for not having an optimal implementation of some function or similar, there may be an alternative way of performing the task that works faster in both browsers, and so lets Quill be usable for large documents cross-platform.

I've had a dig around in the source, but I'm slight over my head. If there's any pointers/guesses anyone more familiar with the code base may have of what/where to look, I'll do some more digging.

**Comment by Aaronik** [comment]
I've been having the same issues with pasting with Firefox. This goes both for pasting HTML and for using `.setContents(delta)`. Performance is not ideal with Chrome, but it's completely unacceptable with Firefox for me, sometimes taking as long as 90 seconds for a 27k character blob on slower machines.

But I found a nice solution that makes firefox lightening fast.

The idea is this: remove the root from the DOM, *then* use `pasteHTML` or `setContents`, then add the root back to the DOM. For me this code looks like this:

```js
const parent = Quill.root.parentElement;
parent.removeChild(Quill.root);
Quill.setContents(delta);
parent.appendChild(Quill.root);
```

Doing that alone has increase performance in Firefox for me by 10 fold!

**Comment by RobAley** [comment]
That looks like a reasonable work-around, I'm not sure if it has any side-
effects. I never got to the bottom of why it was (much) slower on
Firefox (mainly as I didn't have time to get to grips with the code base
and work out what was happening on paste, although your details seem to
indicate it's more likely related to DOM updates than the source of the
data). My project has now moved to a desktop Electron app (for unrelated
reasons) and there are no slowdowns there so I won't be looking it
further. It might be something @jhchen wants to look at in the next
version of Quill.

On Thu, 1 Feb 2018, at 10:14 PM, Aaron Sullivan wrote:
> I've been having the same issues with pasting with Firefox. This goes
> both for pasting HTML and for using .setContents(delta). Performance
> is not ideal with Chrome, but it's completely unacceptable with
> Firefox for me, sometimes taking as long as 90 seconds for a 27k
> character blob on slower machines.> But I found a nice solution that makes firefox lightening fast.


> The idea is this: remove the root from the DOM, *then* use pasteHTML
> or setContents, then add the root back to the DOM. For me this code
> looks like this:> const parent = Quill.root.parentElement;
> parent.removeChild(Quill.root); Quill.setContents(delta);
> parent.appendChild(Quill.root);> Doing that alone has increase performance in Firefox for me by
> 10 fold!> — You are receiving this because you authored the thread. Reply to
> this email directly, view it on GitHub[1], or mute the thread[2].> 


Links:

  1. https://github.com/quilljs/quill/issues/1130#issuecomment-362419974

**Comment by isaaclyman** [comment]
As a shorter example, try copying and pasting the [Wikipedia article on rats](https://en.wikipedia.org/wiki/Rat). The whole article, from the title to the external links, is 22 pages in Microsoft Word and takes ~5 seconds to paste into a Word document, but takes ~26 seconds to paste into the Quill playground in Chrome. Oddly, it only takes ~11 seconds in Firefox, so I guess performance is pretty variable between browsers.

**Comment by timpeteler** [comment]
I recognized serious performance issues in IE11 in conjunction with multiple long code blocks in a quill document. Implementing the mentioned workaround significantly improved the behaviour.

**Comment by benbro** [comment]
Maybe it's possible to run the clipboard [convert function](https://github.com/quilljs/quill/blob/develop/modules/clipboard.js#L81) in a Web Worker and using a [DomParser polyfill](https://stackoverflow.com/a/52087536).

**Comment by mgtest42** [comment]
> const parent = Quill.root.parentElement;
> parent.removeChild(Quill.root);
> Quill.setContents(delta);
> parent.appendChild(Quill.root);

How to use this workaround? I've tried to put it into addMatcher but seems like it breaks everything. In the native on-paste event, the clipboard is empty for me. I suppose clipboard-module catches the event and doesn't propagate it.

**Comment by benbro** [comment]
Tested with v2.0.0-beta.2
https://v2.quilljs.com/standalone/full
Firefox: 30 seconds
Chrome 6 seconds
